File: packages/data-ops/src/contracts/refero.ts
import { z } from "zod";

export const referoReferenceKindSchema = z.enum(["screen", "flow", "style"]);

export const referoPlatformSchema = z.enum(["web", "ios", "android", "unknown"]);

/** Fixed UI Patterns row categories — engine maps Refero screen searches to these. */
export const referoUiPatternCategorySchema = z.enum([
  "onboarding",
  "homepage",
  "pricing",
  "checkout",
  "dashboard",
]);

export const referoSearchRequestSchema = z.object({
  query: z.string().min(1),
  platform: referoPlatformSchema.default("web"),
  limit: z.number().int().positive().max(20).default(8),
  tags: z.array(z.string().min(1)).default([]),
});

export const referoCategorySearchRequestSchema = z.object({
  category: referoUiPatternCategorySchema,
  query: z.string().min(1),
  platform: referoPlatformSchema.default("web"),
  limit: z.number().int().positive().max(20).default(3),
});

/** Rust/Convex JSON uses null for absent optional fields — nullish, not optional-only. */
const referoOptionalTextSchema = z.string().min(1).nullish();

/** Refero MCP may return relative paths or non-standard URLs — do not require absolute URL. */
const referoOptionalUrlSchema = referoOptionalTextSchema;

export const referoReferenceBaseSchema = z.object({
  id: z.string().min(1),
  kind: referoReferenceKindSchema,
  title: z.string().min(1),
  productName: referoOptionalTextSchema,
  productUrl: referoOptionalUrlSchema,
  platform: referoPlatformSchema.default("unknown"),
  sourceUrl: referoOptionalUrlSchema,
  thumbnailUrl: referoOptionalUrlSchema,
  imageUrl: referoOptionalUrlSchema,
  summary: z.string().nullish(),
  tags: z.array(z.string().min(1)).default([]),
  /** Which UI Patterns row this screen was bucketed into (screens only). */
  uiPatternCategory: referoUiPatternCategorySchema.nullish(),
});

export const referoScreenReferenceSchema = referoReferenceBaseSchema.extend({
  kind: z.literal("screen"),
  screenType: referoOptionalTextSchema,
});

export const referoFlowReferenceSchema = referoReferenceBaseSchema.extend({
  kind: z.literal("flow"),
  flowType: referoOptionalTextSchema,
  stepCount: z.number().int().positive().nullish(),
});

export const referoStyleReferenceSchema = referoReferenceBaseSchema.extend({
  kind: z.literal("style"),
  styleType: referoOptionalTextSchema,
});

export const referoReferenceSchema = z.discriminatedUnion("kind", [
  referoScreenReferenceSchema,
  referoFlowReferenceSchema,
  referoStyleReferenceSchema,
]);

export const referoCategorySearchSchema = z.object({
  category: referoUiPatternCategorySchema,
  query: z.string().min(1),
  references: z.array(referoReferenceSchema).default([]),
});

export const referoContextSchema = z.object({
  /** Human-readable summary of all Refero queries in this run. */
  query: z.string().min(1),
  references: z.array(referoReferenceSchema).default([]),
  /** Per screen-type search buckets used to build UI Patterns rows. */
  categorySearches: z.array(referoCategorySearchSchema).default([]),
  fetchedAt: z.number().int().nonnegative(),
});

export type ReferoReferenceKind = z.infer<typeof referoReferenceKindSchema>;
export type ReferoPlatform = z.infer<typeof referoPlatformSchema>;
export type ReferoUiPatternCategory = z.infer<typeof referoUiPatternCategorySchema>;
export type ReferoSearchRequest = z.infer<typeof referoSearchRequestSchema>;
export type ReferoCategorySearchRequest = z.infer<typeof referoCategorySearchRequestSchema>;
export type ReferoScreenReference = z.infer<typeof referoScreenReferenceSchema>;
export type ReferoFlowReference = z.infer<typeof referoFlowReferenceSchema>;
export type ReferoStyleReference = z.infer<typeof referoStyleReferenceSchema>;
export type ReferoReference = z.infer<typeof referoReferenceSchema>;
export type ReferoCategorySearch = z.infer<typeof referoCategorySearchSchema>;
export type ReferoContext = z.infer<typeof referoContextSchema>;

File: apps/stage-engine/src/models/refero.rs
#![allow(dead_code)]
// Contract mirror for Refero MCP references. The adapter is wired into the
// Research workflow after the shared Stage artifact contracts are stable.

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum ReferoReferenceKind {
    Screen,
    Flow,
    Style,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum ReferoPlatform {
    Web,
    Ios,
    Android,
    Unknown,
}

/// UI Patterns row categories shared by Refero app research and Details website research.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum ReferoUiPatternCategory {
    Onboarding,
    Homepage,
    Hero,
    Features,
    #[serde(rename = "social-proof")]
    SocialProof,
    Pricing,
    #[serde(rename = "website-pricing")]
    WebsitePricing,
    Conversion,
    Checkout,
    Dashboard,
}

impl ReferoUiPatternCategory {
    pub fn display_title(self) -> &'static str {
        match self {
            Self::Onboarding => "Onboarding",
            Self::Homepage => "Homepage",
            Self::Hero => "Hero",
            Self::Features => "Features",
            Self::SocialProof => "Social Proof",
            Self::Pricing => "Pricing",
            Self::WebsitePricing => "Pricing",
            Self::Conversion => "Conversion",
            Self::Checkout => "Checkout",
            Self::Dashboard => "Dashboard",
        }
    }

    pub fn row_id(self) -> String {
        format!("ui-patterns-{}", self.as_str())
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Onboarding => "onboarding",
            Self::Homepage => "homepage",
            Self::Hero => "hero",
            Self::Features => "features",
            Self::SocialProof => "social-proof",
            Self::Pricing => "pricing",
            Self::WebsitePricing => "website-pricing",
            Self::Conversion => "conversion",
            Self::Checkout => "checkout",
            Self::Dashboard => "dashboard",
        }
    }

    pub fn all() -> &'static [Self] {
        &[
            Self::Onboarding,
            Self::Homepage,
            Self::Pricing,
            Self::Checkout,
            Self::Dashboard,
        ]
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferoSearchRequest {
    pub query: String,
    pub platform: ReferoPlatform,
    pub limit: u8,
    #[serde(default)]
    pub tags: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferoCategorySearchRequest {
    pub category: ReferoUiPatternCategory,
    pub query: String,
    pub platform: ReferoPlatform,
    pub limit: u8,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferoReference {
    pub id: String,
    pub kind: ReferoReferenceKind,
    pub title: String,
    pub product_name: Option<String>,
    pub product_url: Option<String>,
    pub platform: ReferoPlatform,
    pub source_url: Option<String>,
    pub thumbnail_url: Option<String>,
    pub image_url: Option<String>,
    pub summary: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    pub screen_type: Option<String>,
    pub flow_type: Option<String>,
    pub step_count: Option<u32>,
    pub style_type: Option<String>,
    pub ui_pattern_category: Option<ReferoUiPatternCategory>,
    #[serde(skip)]
    pub raw_image_bytes: Option<Vec<u8>>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferoCategorySearch {
    pub category: ReferoUiPatternCategory,
    pub query: String,
    #[serde(default)]
    pub references: Vec<ReferoReference>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferoContext {
    pub query: String,
    #[serde(default)]
    pub references: Vec<ReferoReference>,
    #[serde(default)]
    pub category_searches: Vec<ReferoCategorySearch>,
    pub fetched_at: u128,
}
